### ISSUE TITLE
Documentation fix for Format autocmds

### ISSUE THREAD

**Issue by tjtrabue** [mhartington/formatter.nvim#32]
Hey everyone,

Let me start off by saying that I absolutely love `formatter.nvim`. I've been dying for a plugin such as this in Neovim for years! I've used it for a few weeks now, and I believe I've found an improved means of using the auto-formatter for certain filetypes.

First, instead of the `BufWritePost` event, we should use `BufWritePre` for auto-formatting buffers for certain filetypes since we want to format the buffer *before* we write it to the filesystem.

Next, it's more flexible in Vim/Neovim to hook into the `FileType` event when designing filetype-specific behavior. Specifying file extensions when creating filetype hooks is a doozy since most filetypes have multiple extensions (i.e., `.pl` and `.plx` for Perl, `.cpp` and `.cxx` for C++, etc.) and sometimes we want to format files that don't even have extensions, such as an executable Bash script. I suggest the following approach for creating auto-format hooks for specific filetypes:

```vim
augroup FormatAutogroup
  autocmd!
  autocmd FileType lua autocmd BufWritePre <buffer> FormatWrite
  autocmd FileType rust autocmd BufWritePre <buffer> FormatWrite
  " Add other filetype hooks below...
augroup end
```

This way we only have to specify the *type* of the file we wish to format, not the extension.

That's all. Thank you again for your hard work! This is a great plugin.

**Comment by lukas-reineke** [comment]
> First, instead of the `BufWritePost` event, we should use `BufWritePre` for auto-formatting buffers for certain filetypes since we want to format the buffer before we write it to the filesystem.

It's not quite as simple as that. This plugin is async, so even if you trigger it before writing to the filesystem, the formatting will be done after it is already written. That is the reason `FormatWrite` exists. `FormatWrite` checks the content of the buffer after the formatting is done, makes sure it is save to overwrite, and then saves the buffer to disc again.
If you were to use `BufWritePre`, there could actually be race conditions in the case that writing is for some reason very slow, and formatting is done before the write happened.
There is a compromise here that the file needs to get written twice, but the performance gain is worth it.

Tldr; `BufWritePost` is correct here


As for the `FileType` autocmd, I agree, that would probably be better 👍

**Comment by tjtrabue** [comment]
Thank you so much for the quick response! I didn't take asynchronous functions into account, I admit. Thank you for the clarification. I'll adjust my own configuration to account for that async performance boost.
